### DISCUSSION TITLE
downloading all dependencies of a package/project

### DISCUSSION THREAD

**Discussion by franzhaas** [pdm-project/pdm#1530] [Ideas]
Dear all,

for me these 2 commands work reasonably well.:
```
pdm export -o requirements.txt
pip download -r requirements.txt
```
thus said, it feels as if there is a better way with pdm...

any hints?

Thanks in advance
Franz

**Comment by frostming** [comment]
It would be easy to implement in a PDM plugin, making use of the API of `unearth`.

**Comment by franzhaas** [comment]
You are right, this was not hard. An initial version is provided here.: https://github.com/franzhaas/pdm-collect-dependencies.

I did not manage to tap into the caching system and did not figure out how to properly output information to the user... but print generally works.

this appears to do the job.

Can someone provide a hint? for accessing the cache and properly output to the user?

Thanks in advance,
Franz

**Reply by pawamoy** [reply]
I believe you can use `project.core.ui.echo` to output messages with colors. You'll find plenty examples in PDM's code base.

**Reply by frostming** [reply]
For caching you can use `with project.environment.get_finder(ignore_compatibility=True) as pf:` the package finder is cache enabled and the session will be closed automatically.

**Comment by franzhaas** [comment]
Thanks a lot, these inputs made it work nicely and with the cache...
